src/smt/theory_str_noodler/theory_str_noodler.cpp

```diff
@@ -1506,6 +1506,43 @@ namespace smt::noodler {
         predicate_replace.insert(r, v.get());
     }
 
+    /**
+     * @brief Handling of str.replace(s,R,t) = v ... s where to replace, R regex what to find, t replacement.
+     * Translates to the following theory axioms:
+     * replace(s,R,t) = v
+     * s \not\in \Sigma*R\Sigma* -> v = s
+     * (s = x.y.z && x \not\in \Sigma*R\Sigma* && y \in R) -> v = x.t.z
+     *
+     * @param e replace_re term
+     */
+    void theory_str_noodler::handle_replace_re(expr *e) {
+        STRACE("str", tout << "handle-replace: " << mk_pp(e, m) << '\n';);
+
+        if(axiomatized_persist_terms.contains(e))
+            return;
```

Comment:
Why is this if check necessary? Why could it be run multiple times?

Reply:
Not completely sure, we have it also in other `handle` functions. I think z3 can send us the same term multiple times. It might be better to do it in `relevant_eh()` instead of in every `handle` function.

Reply:
That makes sense. Thanks for the explanation. Should we move these into the parent handler instead of copying them to every function?

Reply:
It is possible, but it is not that easy. Negated predicates are in the positive form in fact (sign is given by a Boolean flag). Maybe there are some other problems, not sure now. I would keep it as it is.

Reply:
OK. Let us just keep this as it is. Closing this conversation as resolved.